tests/test_PR045_flatten_operation.py

```diff
@@ -68,19 +68,34 @@ def test_flatten_list_offset_array():
     assert awkward1.tolist(array2.flatten()) == [3.3, 4.4, 5.5]
 
 def test_flatten_numpy_array():
-    array = awkward1.layout.NumpyArray(numpy.arange(2*3*5, dtype=numpy.int64).reshape(2, 3, 5))
-    assert awkward1.tolist(array) == [[[ 0,  1,  2,  3,  4], [ 5,  6,  7,  8,  9], [10, 11, 12, 13, 14]],
-                                      [[15, 16, 17, 18, 19], [20, 21, 22, 23, 24], [25, 26, 27, 28, 29]]]
-    assert awkward1.tolist(array.flatten()) == [[0, 1, 2, 3, 4], [5, 6, 7, 8, 9], [10, 11, 12, 13, 14], [15, 16, 17, 18, 19], [20, 21, 22, 23, 24], [25, 26, 27, 28, 29]]
-
-    array2 = awkward1.layout.NumpyArray(numpy.arange(1*6*5, dtype=numpy.int64).reshape(1, 6, 5))
-    assert awkward1.tolist(array2.flatten()) == [[0, 1, 2, 3, 4], [5, 6, 7, 8, 9], [10, 11, 12, 13, 14], [15, 16, 17, 18, 19], [20, 21, 22, 23, 24], [25, 26, 27, 28, 29]]
-
-    array3 = array[:, 1::2, :3]
-    print(awkward1.tolist(array3))
-    assert awkward1.tolist(array3) == [[[5, 6, 7]], [[20, 21, 22]]]
-    assert flatten(awkward1.tolist(array3)) == [[5, 6, 7], [20, 21, 22]]
-    assert awkward1.tolist(array3.flatten()) == [[5, 6, 7], [20, 21, 22]]
+    with pytest.raises(ValueError, match="cannot concatenate non-lists") :
+        array = awkward1.layout.NumpyArray(numpy.arange(2*3*5, dtype=numpy.int64).reshape(2, 3, 5))
+        assert awkward1.tolist(array) == [[[ 0,  1,  2,  3,  4], [ 5,  6,  7,  8,  9], [10, 11, 12, 13, 14]],
+                                          [[15, 16, 17, 18, 19], [20, 21, 22, 23, 24], [25, 26, 27, 28, 29]]]
+        assert flatten(awkward1.tolist(array)) == [[0, 1, 2, 3, 4], [5, 6, 7, 8, 9], [10, 11, 12, 13, 14], [15, 16, 17, 18, 19], [20, 21, 22, 23, 24], [25, 26, 27, 28, 29]]
+        assert awkward1.tolist(array.flatten()) == [[0, 1, 2, 3, 4], [5, 6, 7, 8, 9], [10, 11, 12, 13, 14], [15, 16, 17, 18, 19], [20, 21, 22, 23, 24], [25, 26, 27, 28, 29]]
+        assert flatten(awkward1.tolist(array), 1) == [[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14], [15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29]]
+        assert awkward1.tolist(array.flatten(1)) == [[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14], [15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29]]
+        assert awkward1.tolist(array.flatten(2)) == [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29]
+        # Cannot flaten in axes > shape.size():
+        # assert awkward1.tolist(array.flatten(3)) == 0
```

Comment:
Actually (using the Python-only `flatten` as a guide), it shouldn't be allowed to flatten if `axis >= shape.size()`, not `axis > shape.size()`. The constraint should be a little tighter. This result is due to some technicality.

One rule that should be observed is that `flatten` should reduce the number of dimensions by `1`. Both `array` (above) and `array2` (below) are 3-dimensional arrays, so the output should be 2-dimensional for any legal `axis`. The output of this `array.flatten(axis=2)` is 1-dimensional.

Reply:
The Python-only ```flatten``` fails at ```axis >= shape.size() - 1```.  
Shouldn't flattening in ```axis == 2``` and a **shape (2, 3, 5)** e.g. ```shape.size() == 3``` be something like **shape (10, 3)**:
```python
[[ 0, 1, 2, 3, 4, 15, 16, 17, 18, 19],
[ 5, 6, 7, 8, 9, 20, 21, 22, 23, 24],
[10, 11, 12, 13, 14, 25, 26, 27, 28, 29]]
```